### PR TITLE
[PR 1] Issue-523: Replace go-micro with gRPC framework

### PR DESCRIPTION
**Pull request by a-ajith**
The PR contains the task 1 of the issue 523.
Task 1: Add reusable code to the lib-utilities which supports gRPC implementation

### CODE REVIEW

File: lib-utilities/services/rpcService.go [hunk 1]
```diff
@@ -78,3 +85,124 @@ func InitializeService(serviceName string) error {
 	return nil
 
 }
+
+// gRPC implementation starts here. TODO: remove this comment after the removal of go micro implementation.
+
+type serviceType int
```
Comment (Bharath-KKB) [inline lib-utilities/services/rpcService.go]: What values can this assume ? Can this be an enum ?

File: lib-utilities/services/rpcService.go [hunk 2]
```diff
@@ -78,3 +85,124 @@ func InitializeService(serviceName string) error {
...
+var (
+	// GRPCClientTransportCreds is for establishing the tls secured gRPC client communication
+	GRPCClientTransportCreds credentials.TransportCredentials
+	// GRPCServer is for bringing up gRPC micro services
+	GRPCServer *grpc.Server
+)
+
+// InitializeGRPC initializes the gRPC client transport and server
+// Function returns error at the failure of server or client transport creation
+func InitializeGRPC() error {
```
Comment (Bharath-KKB) [inline lib-utilities/services/rpcService.go]: Are we not implementing Service interface like go micro ?

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
Address comments

File: lib-utilities/services/rpcService.go (hunk 1 above)
Comment (a-ajith) [reply lib-utilities/services/rpcService.go]: yes @Bharath-KKB this is an enum, and the values for this enum are `serverService` & `clientService` mentioned in line 94 and 95

File: lib-utilities/services/rpcService.go (hunk 2 above)
Comment (a-ajith) [reply lib-utilities/services/rpcService.go]: @Bharath-KKB Addressed.

File: lib-utilities/services/rpcService.go [hunk 3]
```diff
@@ -15,66 +15,265 @@
...
 // Service holds the microservice instance
 var Service micro.Service
 
 // InitializeService will initialize a new micro.Service.
 // Service will be initialized here itself, so the Server() and Client()
 // called easily.
-func InitializeService(serviceName string) error {
+func InitializeService(framework frameworkType, serverName string) error {
+	switch framework {
+	case GRPC:
```
Comment (Bharath-KKB) [inline lib-utilities/services/rpcService.go:70]: This is missing TLS config

File: lib-utilities/services/rpcService.go [hunk 4]
```diff
@@ -15,66 +15,265 @@
...
+		if err != nil {
+			return fmt.Errorf("While trying to initiate ODIMService model, got: %v", err)
+		}
+		err = ODIMService.registerService()
+		if err != nil {
+			return fmt.Errorf("While trying to register the service in the registry, got: %v", err)
+		}
+
+	case GoMicro:
+		cer, err := tls.X509KeyPair(
```
Comment (Bharath-KKB) [inline lib-utilities/services/rpcService.go]: can the TLS initialising be part of common code ?

File: lib-utilities/services/rpcService.go [hunk 5]
```diff
@@ -15,66 +15,265 @@
...
 var Service micro.Service
 
 // InitializeService will initialize a new micro.Service.
 // Service will be initialized here itself, so the Server() and Client()
 // called easily.
-func InitializeService(serviceName string) error {
+func InitializeService(framework frameworkType, serverName string) error {
+	switch framework {
+	case GRPC:
+		err := ODIMService.init(serverName)
```
Comment (Bharath-KKB) [inline lib-utilities/services/rpcService.go]: Shouldn't 
 init be Init ?

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
new comments added

File: lib-utilities/services/rpcService.go (hunk 3 above)
Comment (a-ajith) [reply lib-utilities/services/rpcService.go:70]: @Bharath-KKB TLS implementation is moved to init() function.

File: lib-utilities/services/rpcService.go (hunk 5 above)
Comment (a-ajith) [reply lib-utilities/services/rpcService.go]: As the InitializeService function and init() function are both in the same package, we don't have to make it a public function.

File: lib-utilities/services/rpcService.go (hunk 4 above)
Comment (a-ajith) [reply lib-utilities/services/rpcService.go]: addressed. Now the common functions is being used for creating TLS config.

File: lib-utilities/services/rpcService.go (hunk 5 above)
Comment (Bharath-KKB) [reply lib-utilities/services/rpcService.go]: Then that becomes a private interface. Is that what we want ?

Comment (a-ajith) [reply lib-utilities/services/rpcService.go]: Addressed

File: lib-utilities/services/cli.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+// CLIData is for accessing the data passed as the command line argument
+var cliData cliModel
+
+func collectCLIData() {
+	flag.StringVar(&cliData.ClientRequestTimeout, "client_request_timeout", "", "maximum request time which client waits")
+	flag.StringVar(&cliData.Registry, "registry", "", "service registry")
+	flag.StringVar(&cliData.RegistryAddress, "registry_address", "", "address of the registry")
+	flag.StringVar(&cliData.ServerAddress, "server_address", "", "address for the micro service")
+	flag.Parse()
```
Comment (bharath-b-hpe) [inline lib-utilities/services/cli.go]: I think its better to parse all flags in main function, and the values can be stored in variables defined lib-utilities/config package as required.

Comment (a-ajith) [reply lib-utilities/services/cli.go]: Addressed

**Review by bharath-b-hpe — APPROVED** [review]

**Pull request merged**